092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,284 @@
+
+Every configuration related to self-healing is currently disabled in the operator, and we will need to enable users altering some of those properties.
+The setting of these properties will be placed behind a feature gate.
+
+### Feature Gate
+
+The new feature gate will be called `UseSelfHealing`.
+It will be introduced in an alpha state and will be disabled by default.
+
+We are using a feature gate for this feature due to various reasons:
+* We want to allow users to try this feature and test it.
+* We would like to know more about the positive/negatives of this feature by testing it in real deployments.
+* Getting feedback from the community would allow us to understand how we can improve this even more.
+
+### Enabling self-healing in the Kafka custom resource
+
+When the feature gate is enabled, the operator will set the `self.healing.enabled` Cruise Control configuration to `true` and allow the users to set other appropriate `self-healing` related [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations).
+
+Here is an example of how the configured Kafka custom resource could look:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    config:
+      # configuration to enable self-healing for all anomalies
+      self.healing.enabled: true
+```
+
+Where the user wants to enable self-healing for some specific anomalies, they can use configurations like `self.healing.broker.failure.enabled` for broker failure, `self.healing.goal.violation.enabled` for goal violation etc.
+
+### Strimzi Notifier
+
+Cruise Control provides `AnomalyNotifier` interface which has multiple abstract methods on what to do if certain anomalies are detected.
+`SelfHealingNotifier` is the base class that contains the logic of self-healing and override the methods in by implementing the `AnomalyNotifier` interface. 
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDisk Failure`, `alert()` etc.
+The `SelfHealingNotifier` class can then be further used to implement your own notifier
+The other custom notifier offered by Cruise control are also based upon the `SelfHealingNotifier`.
+
+We will be using our own notifier implementation called `StrimziNotifer`(set via the `anomaly.notifier.class` config).
+Strimzi will configure Cruise Control to always use the `StrimziNotifier` to notify regarding the anomalies and take action on them.
+
+For effective alerts and monitoring, we will be making use of Kubernetes events to signal to users what operations Cruise Control is performing automatically.
+The `StrimziNotifier` will override the `alert()` method of the `SelfHealingNotifier` and will trigger and publish events whenever an anomaly is detected.
+These events would publish information regarding the anomaly like the anomalyID, the type of anomaly which was detected etc.
```

Comment:
How is something like the `anomalyID` useful to the user? We do not support the CC UI, so is there some simple way for the users to find more details about this anomaly based on the `anomalyID`?

Reply:
Users can get more information using the `state` endpoint with `executor` substate. The `executor` will show the progress for the anomaly getting resolved(It will also contains the id of anomaly being resolved). they can also use the `state` endpoint with `anomaly_detector` substate to know more about what type of anomaly is detected , when it was detected etc

Reply:
So you expect them to `curl` it?

Reply:
Yes, thats the only way for now to check what Cruise Control is doing during self healing

Reply:
That sounds quite complicated. Do you expect the user would need to do this? Or would they be in 99% of cases be happy with what is in the event itself? The example event does not seem to have muhc detail about the problem that was detected.

Reply:
I guess we can get more information about the anomaly from the notifier.
We cannot get info about the fix because, as per how CC works today, there is no more interaction with the notifier from the anomaly detector after the anomaly fix starts. The notifier has just one goal: to notify and decide for the anomaly detector to fix, delay, ignore the anomaly. Nothing more. It's not called anymore in any way along the process.
Still per our offline discussion, the refused `KafkaAnomaly` custom resource was there to provide a way for the operator to get the anomaly and querying the CC endpoints in order to update such resource with the anomaly fix status in progress.